soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -207,12 +218,123 @@ static int dt_init(const struct device *dev)
 	return intel_adsp_ipc_init(dev);
 }
 
+#ifdef CONFIG_PM_DEVICE
+
+void intel_adsp_ipc_set_resume_handler(const struct device *dev,
+	intel_adsp_ipc_resume_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->resume_fn = fn;
+	api->resume_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+void intel_adsp_ipc_set_suspend_handler(const struct device *dev,
+	intel_adsp_ipc_suspend_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->suspend_fn = fn;
+	api->suspend_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+/**
+ * @brief Manages IPC driver power state change.
+ *
+ * @param dev IPC device.
+ * @param action Power state to be changed to.
+ * @return int Returns 0 on success or optionaly error code from the
+ * registered ipc_power_control_api callbacks.
+ *
+ * @note PM lock is taken at the start of each power transition to prevent concurrent calls
+ * to @ref pm_device_action_run function. Changing Device power state during ISR is not supported.
+ * If IPC Device performs hardware operation and device is busy (what should not happen)
+ * function returns failure. It is API user responsibility to make sure we are not entering
+ * device power transition while device is busy.
+ */
+static int ipc_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	if (k_is_in_isr()) {
+		return -ESRCH;
+	}
```

Comment:
Do you really need this ??

Reply:
Well, you started a subject on TOCTOU race and did not answer my question... Please read my comments above.

Reply:
Nope, it does not solve it. If you have two threads that call `pm_device_action_run()` is possible that both threads pass the check `pm_device_state_is_locked()` and both ends up calling `ipc_pm_action` and in that case you don't know exactly what will happen. That is a problem that device_runtime solves, pm_device_state_lock() was not thought to be used this way (though there is nothing disallowing it). It was simply not thought.

My suggestion is to remove it for now (since is not really protecting) and ensure that you are calling it from a safe context. I will probably change `pm_device_lock()` to return a boolean and fails when the device state is already locked and you can use here it and that will probably be a solution.

Does it work for you ?